Refactor completion for `defaults` command

gaelicWizard opened the pull request:
## Description
I accidentally rewrote the completion functions for the Mac OS X `defaults` command. This version is case-insensitive for all completions, uses `$IFS` to allow for spaces in domain and value names, and works for all my ad hoc testing.

## Motivation and Context
This was an offshoot from another issue and PR and I ended up down the rabbit hole. But now I know way more about _Bash_'s programmable completion!

The goal was to allow case-insensitive completion of domain names, for which an existing PR was pending. That PR had some weirds in it and ended up not actually achieving its goal, so I accidentally fixed it.


## How Has This Been Tested?
Tested locally, and all tests pass.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code follows the code style of this project.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] If I have added a new file, I also added it to ``clean_files.txt`` and formatted it using ``lint_clean_files.sh``.
- [x] I have added tests to cover my changes, and all the new and existing tests pass.

gaelicWizard commented:
I've just noticed that using `-currentHost`/`-host` arguments breaks the completion entirely...

davidpfarrell reviewed (changes requested):
This looks like a legit effort to re-work the completion, and @gaelicWizard seems to be pretty invested it.  So I'm guessing they use it regularly?

However: Regarding the use of `=~ ${cmds// /|}`:

With `cmds` being array, this simply does not work.  `${cmds}` simply resolves to `read` (i.e the first element in the array) - Its the default behavior in bash when referencing an array like a scalar.

I'm happy with the value being an array, but I suggest to use the `_bash-it-array-contains-element` method that we added for this purpose.

Additionally, if you decide to revert the usage of an array:

* if its going to use regex (`=~`)  to match, then [per my previous comment](https://github.com/Bash-it/bash-it/pull/1922#pullrequestreview-735634951) you **must** anchor your values:
_anchor with **^(** and **)$**_
```
=~ ^(${cmds// /|})$
```
This ensures that **only** fully matching words will match.

* Additionally as mentioned, this works:
```
== @(${cmds// /|})
```

This is the technique I would probably go for if I didn't already have `_bash-it-array-contains-element`

Lastly:

Per discussion in #1924 , if  jmah does decide to re-take ownership of the completion, I'm inclined to let them have it (it its original form), vendor it back into bash-it, and submit bug fixes upstream.

gaelicWizard commented:
> Per discussion in #1924 , if jmah does decide to re-take ownership of the completion, I'm inclined to let them have it (it its original form), vendor it back into bash-it, and submit bug fixes upstream.

Definitely.

gaelicWizard commented:
> With `cmds` being array, this simply does not work. `${cmds}` simply resolves to `read` (i.e the first element in the array) - Its the default behavior in bash when referencing an array like a scalar.

I converted it to an array after that and...oops. I was fighting with a different part of it and ended up making lots of arrays instead of wordplitting. Back to the drawing board!

gaelicWizard commented:
> This ensures that **only** fully matching words will match.

We specifically want to match partial words; this is a completion handler! 😜 

> I suggest to use the _bash-it-array-contains-element method that we added for this purpose.

I had not thought to use `_bash-it-array-contains-element`! I absolutely would have if I had thought of it, but now that @jmah might be taking this upstream I feel like I now should make sure not to use it...

I've just reverted the change to making it an array and tested everything myself again. I think I need to write some BATS testing for this. Like I said before, completions are a new area of scripting to me. so many edge cases!

Thanks for helping me with this @davidpfarrell! 

> `== @(${cmds// /|})`
> This is the technique I would probably go for 

I've never seen this syntax before; `@(` `)` is a new one to me. I'm going to try a version using that, but `shellcheck` complains when using `==` with a pattern so I'm not sure how that will play here. `[[` uses unquoted second operands as patterns even with `==`, so it still works, but it *reads* like it's comparing for equality and `=~` is more clear that it's matching a pattern.

---

> pretty invested it. So I'm guessing they use it regularly?

This is literally how I learned _Bash It_ existed 😆

gaelicWizard commented:
> I've never seen this syntax before; `@(` `)` is a new one to me.

It's part of `extglob`! I never use that stuff since it's off by default and I don't like to depend on it. Does _Bash It_ turn that on? Can I rely on it here?

gaelicWizard commented:
I simply cannot make this work with any form of `==`. It only works with `=~` as far as I can wrangle it. 

What am I missing?

gaelicWizard commented:
Ok, this branch is working as-is and ready for merge, using `=~`. I have *alsö* prepared another commit that I can push which requires `extglob` and uses `==`, iff someone tells me that _Bash It_ requires `extglob` and we can rely on it here. (If requiring `extglob` in _Bash It_ is documented somewhere already, please let me know.)

The version using `==` is case-sensitive for the verb (from `$cmds`), which is incorrect but I doubt anyone will ever notice. (Yes, the `defaults` command accepts the verb case-insensitively itself.)

@davidpfarrell, can we require `extglob`?

gaelicWizard commented:
Mwahahaha! It's all over now! I've learned BATS! I'm unstoppable!! Mwahahahahahah!! 🤣🤪

Alsö, I totally just checked in a failing test to my previously-passing PR so...good job me. Anyway, should have that one fixed shortly.

gaelicWizard commented:
The tests take literally over five times longer on the Mac OS X runners than the Ubuntu. 😞

NoahGorny commented on test/completion/defaults.completion.bats:
```diff
@@ -0,0 +1,110 @@
+  function defaults() { echo 'NSGlobalDomain, Bash It'; }
+}
+
+function __check_completion () {
```
I am pretty sure we already have similar setup in other completion tests, can't it be reused?

gaelicWizard replied:
Yes, I literally copied it from bash-it.completion.bats. This function basically made it so I could do these tests. Otherwise I would have spent a week implementing it, or just given up...

I would think that refactoring and moving this function to `test_helper.bash` or somewhere would be out of scope for this PR, though.

NoahGorny replied:
why out of scope? this refactor is simple enough and better than copying the whole function!

NoahGorny reviewed (changes requested):
Hi @gaelicWizard 
Well done on adding tests! Very nice addition!

I left a couple of comments, but it looks very promising :smile:

gaelicWizard commented on test/completion/defaults.completion.bats:
```diff
@@ -0,0 +1,110 @@
+}
+
+@test "completion defaults: ensure that the _defaults function is available" {
+  type -a _defaults &> /dev/null
```
Not if we expect this to move upstream!

gaelicWizard replied:
I would have to load `lib/helpers`, which makes the slow tests even slower, and this is alsö how `bash-it.completion.bats` is. (I copied that file to create this one.) I think that improvements to _Bash It_'s BATS tests are out of scope for this PR.

gaelicWizard commented on test/completion/defaults.completion.bats:
```diff
@@ -0,0 +1,110 @@
+  assert_line -n 0 "$(_known_hosts -a)"
+}
+
+@test "completion defaults: -host teh_host_name - show verbs" {
```
Is not a typo! But I can change it if you prefer 😃

gaelicWizard commented on completion/available/defaults.completion.bash:
```diff
@@ -1,91 +1,122 @@
+
+	#local domains="$( defaults domains )"
+	#local candidates=( $( compgen -W "${domains//, /$'\n'}" | grep -i "^${cur}" ) )
+	local domains="$( defaults domains | sed -e 's/, /^/g' | tr '^' '\n' )"
```
It is possible to have a domain named with a comma and a space...so this will simply never be perfect short of implementing a whole entire new parsing/search engine to enumerate defaults domains.

NoahGorny reviewed (changes requested):
This looks really good, but you still address my refactor comment. You should also add the completion to `clean_files.txt` if you already refactor it

gaelicWizard commented:
Since I've done a 2.0 of this, I'm going to vendor it out anyway.

NoahGorny commented:
> Since I've done a 2.0 of this, I'm going to vendor it out anyway. 

I see, very nice!

You do an amazing work cleaning up stuff and improving the project, thank you so much!

I will try to merge more and more PRs, so we can continue work on new things :smile:

gaelicWizard commented:
I don't know how to make @buhl's `git-vendor` fork work (it puts everything in repo root), and the dangling commit from upstream makes my brain worms riot. What do, @NoahGorny?

NoahGorny commented:
> I don't know how to make @buhl's `git-vendor` fork work (it puts everything in repo root), and the dangling commit from upstream makes my brain worms riot. What do, @NoahGorny? 🥺

Hmm, did you already create a repo for the defaults completion? I'll try to vendor it myself :smile:

NoahGorny commented:
I dont see any repo that you opened @gaelicWizard, so let me know when you open one!

gaelicWizard commented:
> I dont see any repo that you opened @gaelicWizard, so let me know when you open one!

https://twitter.com/fibulaa/status/1438361310379397123

gaelicWizard commented:
@NoahGorny, I've just put up https://github.com/gaelicWizard/bash-progcomp and tagged v2.0.0. I'll do a v2.0.1 and probably a v2.0.2 as I expand the BATS tests.

gaelicWizard commented:
Either way, I figured out the benefits of @buhl's fork! I can amend commits! So, easy enough to just `git mv` and then `--amend`. I'll submit the vendor'd PR today

gaelicWizard commented:
I did not close this. @NoahGorny?

NoahGorny commented:
Are you sure? 
To me it says that you did :O

gaelicWizard commented:
Yeah, it says I did, but I wasn't able to re-open it then...but now I am...!! 🙃

gaelicWizard commented:
> Are you sure?
> To me it says that you did :O

Sorry, I just realized that sounded like I was accusing you. I wasn't! I was asking you to help ♥

davidpfarrell commented:
> Since I've done a 2.0 of this, I'm going to vendor it out anyway.

Pulling thing like this out of bash-it make them more accessible to the community.

You are definitely the best person to own this particular completion - Makes me feel bad I don't use `defaults` more myself :)

Nice work !

NoahGorny commented:
> Either way, I figured out the benefits of @buhl's fork! I can amend commits! So, easy enough to just `git mv` and then `--amend`. I'll submit the vendor'd PR today

Hi @gaelicWizard, let me know if you need any help with the vendoring process- you rock :smiley:

gaelicWizard commented:
@NoahGorny, I'm giving up on `git vendor` for now.

```
git vendor add bash-progcomp https://github.com/gaelicWizard/bash-progcomp v2.0.1 
```

On a freshly cloned repo, this tells me 
```
mv: illegal option -- t
usage: mv [-f | -i | -n] [-v] source target
       mv [-f | -i | -n] [-v] source ... directory
```
and then leaves me with a merge conflict and a dirty tree.

Halp?!

NoahGorny reviewed (approved):
Hi @gaelicWizard 
Please rebase, and then we can merge this :smile:
